runners/spark/src/main/java/org/apache/beam/runners/spark/stateful/StateSpecFunctions.java

```diff
@@ -177,8 +177,6 @@
           Instant sourceWatermark = ((MicrobatchSource.Reader) reader).getWatermark();
           highWatermark = sourceWatermark.isAfter(lowWatermark) ? sourceWatermark : lowWatermark;
 
-          // close and checkpoint reader.
-          reader.close();
```

Comment:
Ideally, we keep the readers open indefinitely, but since the readers can expire and be removed from the cache, we should indeed close them. This can be done by adding a `RemovalListener` to the cache.